File: scripts/gui/priorities.lua
local C2CC = require "scripts.models.combinator"

local priorities = {}

---Parses a Cybersyn 2 signal key string into a SignalID by searching fluid and item prototypes.
function priorities.parse_signal_key(key)
  if not key or type(key) ~= "string" or key == "" then return nil end

  local sig_name = key
  local sig_quality = nil

  -- Clean prefix if present
  if sig_name:sub(1, 6) == "fluid/" then
    sig_name = sig_name:sub(7)
  elseif sig_name:sub(1, 5) == "item/" then
    sig_name = sig_name:sub(6)
  elseif sig_name:sub(1, 8) == "virtual/" then
    sig_name = sig_name:sub(9)
  end

  -- Extract quality if present (e.g. "item-name|rare")
  if sig_name:find("|") then
    local parts = {}
    for part in sig_name:gmatch("[^|]+") do
      table.insert(parts, part)
    end
    sig_name = parts[1]
    if parts[2] and parts[2] ~= "normal" then
      sig_quality = parts[2]
    end
  end

  local p_fluid = prototypes and prototypes.fluid
  local p_item = prototypes and prototypes.item

  local sig_type = nil
  if p_fluid and p_fluid[sig_name] then
    sig_type = "fluid"
  elseif p_item and p_item[sig_name] then
    sig_type = "item"
  else
    return nil
  end

  return {
    type = sig_type,
    name = sig_name,
    quality = sig_quality
  }
end

---Compares two priority lists strictly by unique signal identity (type, name, quality).
---@param old_list table|nil
---@param new_list table|nil
---@return boolean # True if both lists contain identical unique signals
function priorities.are_priorities_equal(old_list, new_list)
  if old_list == new_list then return true end
  if not old_list or not new_list then return false end
  if #old_list ~= #new_list then return false end

  for i = 1, #new_list do
    local a = old_list[i]
    local b = new_list[i]
    if not a or not b or not a.signal or not b.signal then return false end
    if a.signal.name ~= b.signal.name or a.signal.type ~= b.signal.type or a.signal.quality ~= b.signal.quality then
      return false
    end
  end

  return true
end

---Finds the single Cybersyn 2 station associated with this C2CC combinator,
---and queries Cybersyn 2 API to retrieve its inventory IDs.
---@param entity LuaEntity
---@return integer|nil target_stop_unit
---@return integer[] target_inv_ids
function priorities.find_station_for_combinator(entity)
  if not entity or not entity.valid or not remote.interfaces["cybersyn2"] then
    return nil, {}
  end

  local pos = entity.position
  local search_area = {
    { pos.x - 3.5, pos.y - 3.5 },
    { pos.x + 3.5, pos.y + 3.5 }
  }

  -- 1. Find nearby train-stop entity
  local nearby_stops = entity.surface.find_entities_filtered({
    area = search_area,
    type = "train-stop"
  })

  local stop_unit = nearby_stops[1] and nearby_stops[1].unit_number or nil

  -- 2. If no direct train-stop, check straight-rails
  if not stop_unit then
    local nearby_rails = entity.surface.find_entities_filtered({
      area = search_area,
      type = "straight-rail"
    })
    if #nearby_rails > 0 then
      local stops_all = remote.call("cybersyn2", "query", { type = "stops", all = true })
      if stops_all and stops_all.data then
        for _, s in ipairs(stops_all.data) do
          local s_ent = s.entity
          if s_ent and s_ent.valid and s_ent.surface.index == entity.surface.index then
            local dx = s_ent.position.x - entity.position.x
            local dy = s_ent.position.y - entity.position.y
            if (dx * dx + dy * dy) <= 100 then
              stop_unit = s.entity_id or s_ent.unit_number
              break
            end
          end
        end
      end
    end
  end

  if not stop_unit then
    return nil, {}
  end

  -- 3. Query Cybersyn 2 API for this specific station by unit_number
  local stop_res = remote.call("cybersyn2", "query", { type = "stops", unit_numbers = { stop_unit } })
  local stop_obj = stop_res and stop_res.data and stop_res.data[1]

  local target_inv_ids = {}
  local inv_map = {}

  if stop_obj then
    if stop_obj.inventory_id and not inv_map[stop_obj.inventory_id] then
      inv_map[stop_obj.inventory_id] = true
      table.insert(target_inv_ids, stop_obj.inventory_id)
    end

    if stop_obj.shared_inventory_master and not inv_map[stop_obj.shared_inventory_master] then
      inv_map[stop_obj.shared_inventory_master] = true
      table.insert(target_inv_ids, stop_obj.shared_inventory_master)
    end
  end

  return stop_unit, target_inv_ids
end

---Queries Cybersyn 2 for priorities and demand/supply info of unique signals in station inventories near the combinator.
---@param entity LuaEntity Combinator entity.
---@param cached_inv_ids integer[]|nil Optional cached array of station inventory IDs.
---@return table # Array of signal priority stats
function priorities.query_signal_priorities(entity, cached_inv_ids)
  if not entity or not entity.valid or not remote.interfaces["cybersyn2"] then
    return {}
  end

  local comb = C2CC:New(entity)
  local unique_signals_map = {}
  local unique_signals_list = {}

  -- STEP 1: Helper function to register unique signals
  local add_signals_from_dict = function(dict)
    if not dict or type(dict) ~= "table" then return end
    for k in pairs(dict) do
      local parsed = priorities.parse_signal_key(k)
      if parsed and parsed.name and not unique_signals_map[parsed.name] then
        unique_signals_map[parsed.name] = parsed
        table.insert(unique_signals_list, parsed)
      end
    end
  end

  -- STEP 2: Retrieve network mask of the current combinator
  local current_net_sig = comb:GetNetworkSignal()
  local current_net_count = current_net_sig and current_net_sig.Count or nil
  local current_net_name = (current_net_sig and current_net_sig.Signal) and current_net_sig.Signal.name or nil

  -- Helper function to check if order network matches combinator network
  local is_network_matched = function(order_obj, inv)
    if not current_net_sig or not current_net_sig.Signal then
      return true -- Match all networks if no network mask signal is set
    end

    local target_net = order_obj.network or order_obj.network_mask or order_obj.network_id or order_obj.network_flag or
        (inv and (inv.network or inv.network_mask or inv.network_id or inv.network_flag))

    if not target_net then return true end

    if type(target_net) == "number" and type(current_net_count) == "number" and current_net_count ~= 0 then
      return bit32.band(target_net, current_net_count) ~= 0
    elseif type(target_net) == "string" and current_net_name then
      return target_net == current_net_name or target_net:find(current_net_name, 1, true) ~= nil
    end

    return true
  end

  -- STEP 3: Locate target Cybersyn 2 station inventories
  local target_inv_ids = cached_inv_ids
  if not target_inv_ids or #target_inv_ids == 0 then
    local _, resolved_inv_ids = priorities.find_station_for_combinator(entity)
    target_inv_ids = resolved_inv_ids
  end

  -- Query target station inventories and extract unique items/fluids
  if target_inv_ids and #target_inv_ids > 0 then
    local inv_res = remote.call("cybersyn2", "query", { type = "inventories", ids = target_inv_ids })
    if inv_res and inv_res.data then
      for _, inv in ipairs(inv_res.data) do
        add_signals_from_dict(inv.inventory)
        add_signals_from_dict(inv.inflow)
        add_signals_from_dict(inv.outflow)

        if inv.orders then
          for _, order_obj in ipairs(inv.orders) do
            add_signals_from_dict(order_obj.provides)
            add_signals_from_dict(order_obj.requests)
            add_signals_from_dict(order_obj.requested_fluids)
          end
        end
      end
    end
  end

  -- Return empty result if no signals found
  if #unique_signals_list == 0 then
    return {}
  end

  -- STEP 4: Initialize structured priority stats container for each signal
  local signal_stats = {}
  for _, sig in ipairs(unique_signals_list) do
    signal_stats[sig.name] = {
      signal = sig,
      req_min = nil,
      req_max = nil,
      req_found = false,
      sup_min = nil,
      sup_max = nil,
      sup_found = false,
    }
  end

  -- STEP 5: Query all Cybersyn 2 orders and calculate Min/Max priorities for Request and Supply
  local stops_res = remote.call("cybersyn2", "query", { type = "stops", all = true })
  if stops_res and stops_res.data then
    local inv_ids = {}
    for _, stop in ipairs(stops_res.data) do
      if stop.inventory_id then
        table.insert(inv_ids, stop.inventory_id)
      end
      if stop.shared_inventory_master then
        table.insert(inv_ids, stop.shared_inventory_master)
      end
    end

    if #inv_ids > 0 then
      local inv_res = remote.call("cybersyn2", "query", { type = "inventories", ids = inv_ids })
      if inv_res and inv_res.data then
        local process_order_dict = function(dict, is_request, prio)
          if not dict or type(dict) ~= "table" then return end
          for k in pairs(dict) do
            local parsed = priorities.parse_signal_key(k)
            if parsed and parsed.name and signal_stats[parsed.name] then
              local stat = signal_stats[parsed.name]
              if is_request then
                stat.req_found = true
                if not stat.req_max or prio > stat.req_max then stat.req_max = prio end
                if not stat.req_min or prio < stat.req_min then stat.req_min = prio end
              else
                stat.sup_found = true
                if not stat.sup_max or prio > stat.sup_max then stat.sup_max = prio end
                if not stat.sup_min or prio < stat.sup_min then stat.sup_min = prio end
              end
            end
          end
        end

        for _, inv in ipairs(inv_res.data) do
          if inv.orders then
            for _, order_obj in ipairs(inv.orders) do
              if is_network_matched(order_obj, inv) then
                local prio = order_obj.priority or 0
                process_order_dict(order_obj.requests, true, prio)
                process_order_dict(order_obj.requested_fluids, true, prio)
                process_order_dict(order_obj.provides, false, prio)
              end
            end
          end
        end
      end
    end
  end

  -- STEP 6: Build final array of priority statistics for GUI rendering
  local result_list = {}
  for _, sig in ipairs(unique_signals_list) do
    local stat = signal_stats[sig.name]
    table.insert(result_list, {
      signal = sig,
      req_min = stat and stat.req_min or nil,
      req_max = stat and stat.req_max or nil,
      req_found = stat and stat.req_found or false,
      sup_min = stat and stat.sup_min or nil,
      sup_max = stat and stat.sup_max or nil,
      sup_found = stat and stat.sup_found or false,
    })
  end

  return result_list
end

return priorities

File: src/scripts/gui/priorities.ts
import { Combinator } from '../models/combinator';
import { strace } from 'fcore/utils/strace';
import { modPrefix } from '../constants';
import type { SignalID, LuaEntity, BoundingBox } from 'factorio:runtime';

/**
 * Resolves active Cybersyn 2 remote interface.
 */
export function getCybersynInterface(): string | undefined {
  if (typeof remote === 'undefined' || !remote.interfaces) return undefined;
  if (remote.interfaces['cybersyn2']) return 'cybersyn2';
  return undefined;
}

const signalKeyCache: Record<string, SignalID | undefined> = {};

/**
 * Parses signal keys from Cybersyn 2 into structured SignalID with memoization.
 * Format: "name" (normal quality) or "name|quality".
 */
export function parseSignalKey(key: string): SignalID | undefined {
  if (!key || typeof key !== 'string' || key === '') return undefined;
  if (signalKeyCache[key] !== undefined) {
    return signalKeyCache[key];
  }

  let sigName = key;
  let sigQuality: string | undefined = undefined;

  const pipeIndex = sigName.indexOf('|');
  if (pipeIndex !== -1) {
    const q = sigName.substring(pipeIndex + 1);
    sigName = sigName.substring(0, pipeIndex);
    if (q !== 'normal') {
      sigQuality = q;
    }
  }

  let sigType:
    | 'item'
    | 'fluid'
    | 'virtual'
    | 'quality'
    | 'entity'
    | 'recipe'
    | 'space-location'
    | 'asteroid-chunk'
    | undefined = undefined;

  if (prototypes.item && prototypes.item[sigName] !== undefined) {
    sigType = 'item';
  } else if (prototypes.fluid && prototypes.fluid[sigName] !== undefined) {
    sigType = 'fluid';
  } else if (prototypes.virtual_signal && prototypes.virtual_signal[sigName] !== undefined) {
    sigType = 'virtual';
  } else if (prototypes.quality && prototypes.quality[sigName] !== undefined) {
    sigType = 'quality';
  } else if (prototypes.entity && prototypes.entity[sigName] !== undefined) {
    sigType = 'entity';
  } else if (prototypes.recipe && prototypes.recipe[sigName] !== undefined) {
    sigType = 'recipe';
  } else if (prototypes.space_location && prototypes.space_location[sigName] !== undefined) {
    sigType = 'space-location';
  } else if (prototypes.asteroid_chunk && prototypes.asteroid_chunk[sigName] !== undefined) {
    sigType = 'asteroid-chunk';
  } else {
    signalKeyCache[key] = undefined;
    return undefined;
  }

  const result: SignalID = {
    type: sigType as any,
    name: sigName,
    quality: sigQuality as any,
  };
  signalKeyCache[key] = result;
  return result;
}

/**
 * Finds the Cybersyn station associated with the combinator by nearby train-stop or rails.
 */
export function findStationForCombinator(
  entity: LuaEntity,
): LuaMultiReturn<[number | undefined, number[]]> {
  const csInterface = getCybersynInterface();
  if (!entity || !entity.valid || !csInterface) {
    strace.trace(
      modPrefix,
      'find_station_skip',
      'valid',
      entity?.valid,
      'csInterface',
      csInterface,
    );
    return $multi(undefined, []);
  }

  const pos = entity.position as any;
  const px = (pos.x ?? pos[0]) as number;
  const py = (pos.y ?? pos[1]) as number;

  const searchArea: BoundingBox = {
    left_top: { x: px - 3.5, y: py - 3.5 },
    right_bottom: { x: px + 3.5, y: py + 3.5 },
  };

  // 1. Direct train-stop in 3.5 tile radius
  const nearbyStops = entity.surface.find_entities_filtered({
    area: searchArea,
    type: 'train-stop',
  });

  let stopUnit = nearbyStops[0]?.unit_number;

  // 2. If no direct train-stop, check nearby rails and match Cybersyn stops
  if (!stopUnit) {
    const nearbyRails = entity.surface.find_entities_filtered({
      area: searchArea,
      type: [
        'legacy-straight-rail',
        'straight-rail-horizontal',
        'straight-rail-vertical',
        'half-diagonal-rail',
        'curved-rail-a',
        'curved-rail-b',
        'rail-ramp',
        'elevated-straight-rail',
        'elevated-curved-rail-a',
        'elevated-curved-rail-b',
        'elevated-half-diagonal-rail',
      ] as any,
    });
    if (nearbyRails.length > 0) {
      const stopsAll = remote.call(csInterface, 'query', { type: 'stops', all: true }) as any;
      if (stopsAll && stopsAll.data) {
        for (const s of stopsAll.data) {
          const sEnt = s.entity;
          if (sEnt && sEnt.valid && sEnt.surface.index === entity.surface.index) {
            const sPos = sEnt.position as any;
            const sx = (sPos.x ?? sPos[0]) as number;
            const sy = (sPos.y ?? sPos[1]) as number;
            const dx = sx - px;
            const dy = sy - py;
            if (dx * dx + dy * dy <= 100) {
              stopUnit = s.entity_id || sEnt.unit_number;
              break;
            }
          }
        }
      }
    }
  }

  if (!stopUnit) {
    strace.trace(
      modPrefix,
      'find_station_none',
      'combinator',
      entity.unit_number,
      'pos_x',
      px,
      'pos_y',
      py,
    );
    return $multi(undefined, []);
  }

  const stopRes = remote.call(csInterface, 'query', {
    type: 'stops',
    unit_numbers: [stopUnit],
  }) as any;
  const stopList = stopRes && stopRes.data;
  let stopObj: any = undefined;
  if (stopList) {
    for (const s of stopList) {
      if (s) {
        stopObj = s;
        break;
      }
    }
  }

  const targetInvIds: number[] = [];
  const invMap: Record<number, boolean> = {};

  if (stopObj) {
    const invId = stopObj.inventory_id || stopObj.created_inventory_id;
    if (invId && !invMap[invId]) {
      invMap[invId] = true;
      targetInvIds.push(invId);
    }
    if (stopObj.shared_inventory_master) {
      const masterRes = remote.call(csInterface, 'query', {
        type: 'stops',
        ids: [stopObj.shared_inventory_master],
      }) as any;
      if (masterRes && masterRes.data) {
        for (const m of masterRes.data) {
          const mInvId = m && (m.inventory_id || m.created_inventory_id);
          if (mInvId && !invMap[mInvId]) {
            invMap[mInvId] = true;
            targetInvIds.push(mInvId);
          }
        }
      }
    }
  }

  strace.trace(
    modPrefix,
    'find_station_success',
    'combinator',
    entity.unit_number,
    'stop_unit',
    stopUnit,
    'inventories',
    targetInvIds.length,
  );

  return $multi(stopUnit, targetInvIds);
}

let cachedAllInvData: any[] | undefined = undefined;
let cachedAllInvTick = 0;

function getAllWorldInventories(csInterface: string): any[] | undefined {
  const currentTick = typeof game !== 'undefined' && game ? game.tick : 0;
  if (cachedAllInvData !== undefined && currentTick - cachedAllInvTick < 60) {
    return cachedAllInvData;
  }

  const stopsRes = remote.call(csInterface, 'query', { type: 'stops', all: true }) as any;
  if (!stopsRes || !stopsRes.data) {
    return undefined;
  }

  const invIds: number[] = [];
  const invSeen: Record<number, boolean> = {};
  for (const stop of stopsRes.data) {
    const invId = stop.inventory_id || stop.created_inventory_id;
    if (invId && !invSeen[invId]) {
      invSeen[invId] = true;
      invIds.push(invId);
    }
  }

  if (invIds.length === 0) return undefined;

  const allInvRes = remote.call(csInterface, 'query', {
    type: 'inventories',
    ids: invIds,
  }) as any;

  if (allInvRes && allInvRes.data) {
    cachedAllInvData = allInvRes.data;
    cachedAllInvTick = currentTick;
    return cachedAllInvData;
  }

  return undefined;
}

/**
 * Queries Cybersyn 2 for global priorities and demand/supply statistics for unique signals.
 */
export function querySignalPriorities(
  entity: LuaEntity,
  cachedInvIds?: number[],
): SignalPriorityStat[] {
  const csInterface = getCybersynInterface();
  if (!entity || !entity.valid || !csInterface) {
    strace.trace(modPrefix, 'query_skip', 'valid', entity?.valid, 'csInterface', csInterface);
    return [];
  }

  // STEP 1: Locate target Cybersyn 2 station inventories first (early exit if no station)
  let targetInvIds = cachedInvIds;
  if (!targetInvIds || targetInvIds.length === 0) {
    const [_, resolvedInvIds] = findStationForCombinator(entity);
    targetInvIds = resolvedInvIds;
  }

  if (!targetInvIds || targetInvIds.length === 0) {
    strace.trace(modPrefix, 'no_station_inventories', 'combinator', entity.unit_number);
    return [];
  }

  const invRes = remote.call(csInterface, 'query', {
    type: 'inventories',
    ids: targetInvIds,
  }) as any;

  if (!invRes || !invRes.data || invRes.data.length === 0) {
    strace.trace(modPrefix, 'no_inventory_data', 'combinator', entity.unit_number);
    return [];
  }

  // STEP 2: Collect unique item/fluid signals from station orders/inventory
  const uniqueSignalsMap: Record<string, SignalID> = {};
  const uniqueSignalsList: SignalID[] = [];

  const addSignalsFromDict = (dict: any) => {
    if (!dict || typeof dict !== 'object') return;
    for (const [k] of pairs(dict as Record<string, number>)) {
      const parsed = parseSignalKey(k as string);
      if (parsed && parsed.name && !uniqueSignalsMap[parsed.name]) {
        uniqueSignalsMap[parsed.name] = parsed;
        uniqueSignalsList.push(parsed);
      }
    }
  };

  for (const inv of invRes.data) {
    addSignalsFromDict(inv.inventory);
    addSignalsFromDict(inv.inflow);
    addSignalsFromDict(inv.outflow);

    if (inv.orders) {
      for (const orderObj of inv.orders) {
        addSignalsFromDict(orderObj.provides);
        addSignalsFromDict(orderObj.requests);
        addSignalsFromDict(orderObj.requested_fluids);
      }
    }
  }

  // Also include signals configured in the combinator's custom sections
  const comb = new Combinator(entity);
  const cb = comb.getControlBehavior();
  if (cb && cb.valid) {
    const totalSections = cb.sections_count;
    for (let secIdx = 3; secIdx <= totalSections; secIdx++) {
      const sec = cb.get_section(secIdx);
      if (sec && sec.valid) {
        const filters = sec.filters;
        if (filters !== undefined) {
          for (const f of filters) {
            if (f && f.value) {
              const val = f.value;
              const sName = typeof val === 'string' ? val : val.name;
              const sType = typeof val === 'string' ? 'item' : val.type || 'item';
              const sQuality = typeof val === 'object' ? val.quality : undefined;
              if (sName && !uniqueSignalsMap[sName]) {
                const sig: SignalID = { type: sType as any, name: sName, quality: sQuality };
                uniqueSignalsMap[sName] = sig;
                uniqueSignalsList.push(sig);
              }
            }
          }
        }
      }
    }
  }

  strace.trace(
    modPrefix,
    'station_signals',
    'combinator',
    entity.unit_number,
    'signals_found',
    uniqueSignalsList.length,
    'inventories',
    targetInvIds.length,
  );

  if (uniqueSignalsList.length === 0) return [];

  // STEP 3: Retrieve network mask of the current combinator
  const currentNetSig = comb.getNetworkSignal();
  const currentNetCount = currentNetSig?.count;
  const currentNetName = currentNetSig?.signal?.name;

  const isNetworkMatched = (orderObj: any, inv: any): boolean => {
    if (!currentNetSig || !currentNetSig.signal) return true;
    const targetNet =
      orderObj.network ||
      orderObj.network_mask ||
      orderObj.network_id ||
      orderObj.network_flag ||
      (inv && (inv.network || inv.network_mask || inv.network_id || inv.network_flag));

    if (!targetNet) return true;

    if (
      typeof targetNet === 'number' &&
      typeof currentNetCount === 'number' &&
      currentNetCount !== 0
    ) {
      return (targetNet & currentNetCount) !== 0;
    } else if (typeof targetNet === 'string' && currentNetName) {
      return targetNet === currentNetName || (targetNet as string).indexOf(currentNetName) !== -1;
    }

    return true;
  };

  // STEP 4: Initialize structured priority stats container for each signal
  const signalStats: Record<string, SignalPriorityStat> = {};
  for (const sig of uniqueSignalsList) {
    if (sig.name) {
      signalStats[sig.name] = {
        signal: sig,
        reqFound: false,
        supFound: false,
      };
    }
  }

  // STEP 5: Query all Cybersyn orders in the world and compute Min/Max priorities
  const allInventories = getAllWorldInventories(csInterface);
  if (allInventories && allInventories.length > 0) {
    const processOrderDict = (dict: any, isRequest: boolean, prio: number) => {
      if (!dict || typeof dict !== 'object') return;
      for (const [k] of pairs(dict as Record<string, number>)) {
        const keyStr = k as string;
        let stat = signalStats[keyStr];
        if (stat === undefined) {
          const pipeIdx = keyStr.indexOf('|');
          if (pipeIdx !== -1) {
            stat = signalStats[keyStr.substring(0, pipeIdx)];
          }
        }

        if (stat !== undefined) {
          if (isRequest) {
            stat.reqFound = true;
            if (stat.reqMax === undefined || prio > stat.reqMax) stat.reqMax = prio;
            if (stat.reqMin === undefined || prio < stat.reqMin) stat.reqMin = prio;
          } else {
            stat.supFound = true;
            if (stat.supMax === undefined || prio > stat.supMax) stat.supMax = prio;
            if (stat.supMin === undefined || prio < stat.supMin) stat.supMin = prio;
          }
        }
      }
    };

    for (const inv of allInventories) {
      if (inv.orders) {
        for (const orderObj of inv.orders) {
          if (isNetworkMatched(orderObj, inv)) {
            const prio = tonumber(orderObj.priority) || 0;
            processOrderDict(orderObj.requests, true, prio);
            processOrderDict(orderObj.requested_fluids, true, prio);
            processOrderDict(orderObj.provides, false, prio);
          }
        }
      }
    }
  }

  // STEP 6: Build final array of priority statistics for GUI rendering
  const resultList: SignalPriorityStat[] = [];
  for (const sig of uniqueSignalsList) {
    if (sig.name && signalStats[sig.name]) {
      resultList.push(signalStats[sig.name]);
    }
  }

  strace.trace(
    modPrefix,
    'query_result',
    'combinator',
    entity.unit_number,
    'stats_count',
    resultList.length,
  );

  return resultList;
}
